ethers-core/src/utils/units.rs

```diff
@@ -1,63 +1,122 @@
 use super::ConversionError;
+use std::{convert::TryFrom, fmt, str::FromStr};
 
 /// Common Ethereum unit types.
+#[derive(Clone, Copy, Debug, PartialEq, Eq, PartialOrd, Ord)]
 pub enum Units {
-    /// Ether corresponds to 1e18 Wei
-    Ether,
-    /// Gwei corresponds to 1e9 Wei
-    Gwei,
-    /// Wei corresponds to 1 Wei
+    /// Wei is equivalent to 1 wei.
     Wei,
-    /// Use this for other less frequent unit sizes
+    /// Gwei is equivalent to 1e9 wei.
+    Gwei,
+    /// Ether is equivalent to 1e18 wei.
+    Ether,
+    /// Other less frequent unit sizes, equivalent to 1e{0} wei.
     Other(u32),
 }
 
-impl Units {
-    pub fn as_num(&self) -> u32 {
-        match self {
-            Units::Ether => 18,
-            Units::Gwei => 9,
-            Units::Wei => 0,
-            Units::Other(inner) => *inner,
-        }
+impl fmt::Display for Units {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.pad(self.as_num().to_string().as_str())
     }
 }
 
-use std::convert::TryFrom;
+/* ----------------------------------------- Into<Units> ---------------------------------------- */
 
 impl TryFrom<u32> for Units {
     type Error = ConversionError;
 
-    fn try_from(src: u32) -> Result<Self, Self::Error> {
-        Ok(Units::Other(src))
+    fn try_from(value: u32) -> Result<Self, Self::Error> {
+        Ok(Units::Other(value))
     }
 }
 
 impl TryFrom<i32> for Units {
     type Error = ConversionError;
 
-    fn try_from(src: i32) -> Result<Self, Self::Error> {
-        Ok(Units::Other(src as u32))
+    fn try_from(value: i32) -> Result<Self, Self::Error> {
+        Ok(Units::Other(value as u32))
     }
 }
 
 impl TryFrom<usize> for Units {
     type Error = ConversionError;
 
-    fn try_from(src: usize) -> Result<Self, Self::Error> {
-        Ok(Units::Other(src as u32))
+    fn try_from(value: usize) -> Result<Self, Self::Error> {
+        Ok(Units::Other(value as u32))
     }
 }
 
-impl std::convert::TryFrom<&str> for Units {
+impl TryFrom<String> for Units {
     type Error = ConversionError;
 
-    fn try_from(src: &str) -> Result<Self, Self::Error> {
-        Ok(match src.to_lowercase().as_str() {
+    fn try_from(value: String) -> Result<Self, Self::Error> {
+        Self::from_str(&value)
+    }
+}
+
+impl TryFrom<&str> for Units {
+    type Error = ConversionError;
+
+    fn try_from(value: &str) -> Result<Self, Self::Error> {
+        Self::from_str(value)
+    }
+}
```

Comment:
I can't remember if there was an issue with `TryFrom<T:AsRef<str>>` in general, so not sure if this would work.

if `TryFrom<T:AsRef<str>>` is not possible we should also add `impl<'a> TryFrom<&'a String>`

Reply:
Can't figure out how to bound `T` to `AsRef<str>`, getting T not found or conflicting implementations of TryFrom

Reply:
yeh there's apparently an issue with blanket tryfrom impls which doesn't allow this